### DISCUSSION TITLE
[panic]在某些网络比较差的场景下，quic_read函数出现panic，帮忙看下是否有遇到过

### DISCUSSION THREAD

**Discussion by sharewzy** [Tencent/tquic#432] [General]
[1121.zip](https://github.com/user-attachments/files/17845218/1121.zip)

网络比较差的情况下会出现以下crash异常。

11-21 14:41:45.154 29267 29508 F libc    : Fatal signal 11 (SIGSEGV), code -6 (SI_TKILL) in tid 29508 (NGairChannel), pid 29267 (atwall.cloudapp)
11-21 14:41:45.510 29815 29815 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
11-21 14:41:45.510 29815 29815 F DEBUG   : Build fingerprint: 'Adayo/salvator/salvator:9/PQ2A.190405.003-RELtest-r8a7795/Jenkins10311754:userdebug/test-keys'
11-21 14:41:45.510 29815 29815 F DEBUG   : Revision: '2.0'
11-21 14:41:45.510 29815 29815 F DEBUG   : ABI: 'arm64'
11-21 14:41:45.510 29815 29815 F DEBUG   : pid: 29267, tid: 29508, name: NGairChannel  >>> com.greatwall.cloudapp <<<
11-21 14:41:45.510 29815 29815 F DEBUG   : signal 11 (SIGSEGV), code -6 (SI_TKILL), fault addr --------
11-21 14:41:45.510 29815 29815 F DEBUG   :     x0  0000f6699bc23ae0  x1  0000f669b21aeb62  x2  0000000000000000  x3  0000000000000000
11-21 14:41:45.510 29815 29815 F DEBUG   :     x4  0000f669a7cf17e0  x5  0000f6699ed9dacc  x6  695f6362696c5f5f  x7  000000002cdf996e
11-21 14:41:45.510 29815 29815 F DEBUG   :     x8  0000000000000000  x9  0000f669cd6e2ca8  x10 0000000000000400  x11 00000000ffffffff
11-21 14:41:45.511 29815 29815 F DEBUG   :     x12 0000f669ae66d280  x13 0000000000000000  x14 000000000000000c  x15 0000000000000000
11-21 14:41:45.511 29815 29815 F DEBUG   :     x16 0000f669b21c6bc8  x17 0000f66a51edf100  x18 0000f6699bc10718  x19 0000f669bf67b8a4
11-21 14:41:45.511 29815 29815 F DEBUG   :     x20 0000000000000400  x21 0000f6699eab5300  x22 0000f66a51ec1000  x23 0000f669b21c7db8
11-21 14:41:45.511 29815 29815 F DEBUG   :     x24 0000f669b21fef18  x25 0000f669a780b588  x26 0000f669b21ff030  x27 0000f669ad7ab544
11-21 14:41:45.511 29815 29815 F DEBUG   :     x28 0000f669addf10a0  x29 0000f669bf67b880
11-21 14:41:45.511 29815 29815 F DEBUG   :     sp  0000f669bf67b860  lr  0000f669b21955c4  pc  0000f669b21955d8
11-21 14:41:45.523 29815 29815 F DEBUG   :
11-21 14:41:45.523 29815 29815 F DEBUG   : backtrace:
11-21 14:41:45.524 29815 29815 F DEBUG   :     #00 pc 000000000000d5d8  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x2621000) (GetAbortMessageBeforeQ+164)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #01 pc 000000000000b13c  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x2621000) (dumpAbortMessage+120)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #02 pc 0000000000008f08  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x2621000)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #03 pc 0000000000000664  [vdso:0000f66a53157000]
11-21 14:41:45.524 29815 29815 F DEBUG   :     #04 pc 0000000000021ab8  /system/lib64/libc.so (abort+120)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #05 pc 0000000000979c58  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (std::sys::pal::unix::abort_internal::h22ff110bbc1699b8+4)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #06 pc 0000000000970f4c  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (rust_panic+88)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #07 pc 0000000000970d20  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (std::panicking::rust_panic_with_hook::h934a8048af49d358+660)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #08 pc 0000000000970a28  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #09 pc 000000000096e344  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #10 pc 00000000009707ec  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (rust_begin_unwind+44)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #11 pc 0000000000993724  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (core::panicking::panic_fmt::ha067cbb55934613d+44)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #12 pc 00000000009937a4  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (core::panicking::panic::h7675799b3abc9261+48)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #13 pc 0000000000704e88  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #14 pc 00000000006fa458  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #15 pc 0000000000646c28  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (tquic::connection::Connection::stream_read::h77f509fb12033833+148)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #16 pc 00000000005fb5c0  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (quic_stream_read+408)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #17 pc 00000000005e1398  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairChannel::recvExcatllyLength(void*, int)+436)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #18 pc 00000000005e19a0  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairChannel::recvMsgBody()+88)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #19 pc 00000000005e1810  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairChannel::recvOneMessage()+132)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #20 pc 00000000005e1a3c  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairChannel::recvMessages()+32)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #21 pc 00000000005dabe8  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairClient::notifyStreamReadable(int)+56)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #22 pc 00000000005daba0  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (client_on_stream_readable(void*, quic_conn_t*, unsigned long)+384)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #23 pc 00000000005fc2cc  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (_$LT$tquic..ffi..TransportHandler$u20$as$u20$tquic..TransportHandler$GT$::on_stream_readable::h6658f8158140c6ac+132)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #24 pc 0000000000666ef4  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.524 29815 29815 F DEBUG   :     #25 pc 0000000000666408  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (tquic::endpoint::Endpoint::process_connections::hf2d34023b1d473b4+564)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #26 pc 00000000005f98dc  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (quic_endpoint_process_connections+20)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #27 pc 00000000005d7f50  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairClient::processConnection()+28)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #28 pc 00000000005d8a78  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #29 pc 00000000005f27d8  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (ev_invoke_pending+184)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #30 pc 00000000005f2cb8  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (ev_run+1224)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #31 pc 00000000005e6890  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #32 pc 00000000005e6860  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairLooper::startLooper()+36)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #33 pc 00000000005d7fe0  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (GairClient::startGairClientLooper()+24)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #34 pc 00000000005d7b24  /data/app/com.greatwall.cloudapp-vP1WdB-5yBPcTGPmpRDkcA==/base.apk (offset 0x28b6000) (gairClientThread(void*)+1504)
11-21 14:41:45.525 29815 29815 F DEBUG   :     #35 pc 0000000000083114  /system/lib64/libc.so (__pthread_start(void*)+36)
11-21 14:41:45.526 29815 29815 F DEBUG   :     #36 pc 00000000000233bc  /system/lib64/libc.so (__start_thread+68)

**Comment by iyangsj** [comment]
目前没有遇到类似问题。请问使用的tquic协议库版本是多少？每次偶发出现panic的调用栈是否相同？是否有复现的方式

**Reply by iyangsj** [reply]
从日志中调用栈初步分析，问题由于访问内存非法导致。
发生panic最顶层的函数调用是 [tquic::connection::Connection::stream_read()](https://github.com/Tencent/tquic/blob/develop/src/connection/connection.rs#L3846C1-L3865C1)，而不是stream_read()函数进一步调用的函数，因此可以先排除line 3849/3850/3852/3857；而函数中剩余行似乎不可能发生core，除非内存被破坏 （例如Connection对象已被释放）

<img width="845" alt="image" src="https://github.com/user-attachments/assets/9d32a18a-9c1a-4f5f-95ce-6e14d602ebf4">
